src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -15,102 +15,568 @@
+            .setRequestConfigCallback(this::customizeRequestConfig)
+    );
+    this.bulkProcessor = BulkProcessor
+        .builder((req, lis) -> client.bulkAsync(req, RequestOptions.DEFAULT, lis), buildListener())
+        .setBulkActions(config.batchSize())
+        .setConcurrentRequests(config.maxInFlightRequests() - 1) // 0 = no concurrent requests
+        .setFlushInterval(TimeValue.timeValueMillis(config.lingerMs()))
+        .setBackoffPolicy(
+            BackoffPolicy.exponentialBackoff(
+                TimeValue.timeValueMillis(config.retryBackoffMs()),
+                config.maxRetries()
+            )
+        )
+        .build();
+  }
+
+  /**
+   * Returns the underlying Elasticsearch client.
+   *
+   * @return the underlying RestHighLevelClient
+   */
+  public RestHighLevelClient client() {
+    return client;
+  }
+
+  /**
+   * Closes the ElasticsearchClient.
+   *
+   * @throws ConnectException if all of the records fail to flush before the timeout.
+   */
+  public void close() {
+    try {
+      if (!bulkProcessor.awaitClose(1, TimeUnit.MINUTES)) {
+        throw new ConnectException("Failed to process all outstanding requests in time.");
+      }
+    } catch (InterruptedException e) {
+      log.error("Bulk processor close was interrupted.", e);
+    }
+  }
+
+  /**
+   * Creates an index. Will not recreate the index if it already exists.
+   *
+   * @param index the index to create
+   */
+  public void createIndex(String index) {
+    if (indexExists(index)) {
+      return;
+    }
+
+    CreateIndexRequest request = new CreateIndexRequest(index);
+    callWithRetries(
+        "create index " + index,
+        () -> {
+          try {
+            client.indices().create(request, RequestOptions.DEFAULT);
+          } catch (ElasticsearchStatusException | IOException e) {
+            if (!e.getMessage().contains(RESOURCE_ALREADY_EXISTS_EXCEPTION)) {
+              throw e;
+            }
+          }
+
+          return null;
+        }
+    );
+  }
+
+  /**
+   * Creates a mapping for the given index and schema.
+   *
+   * @param index the index to create the mapping for
+   * @param schema the schema to map
+   */
+  public void createMapping(String index, Schema schema) {
+    PutMappingRequest request = new PutMappingRequest(index).source(Mapping.buildMapping(schema));
+    callWithRetries(
+        "create mapping",
+        () -> client.indices().putMapping(request, RequestOptions.DEFAULT)
+    );
+  }
+
+  /**
+   * Flushes any buffered records.
+   */
+  public void flush() {
+    bulkProcessor.flush();
+  }
+
+  /**
+   * Checks whether the index already has a mapping or not.
+   * @param index the index to check
+   * @return true if a mapping exists, false if it does not
+   */
+  public boolean hasMapping(String index) {
+    MappingMetaData mapping = getMapping(index);
+    return mapping != null && !mapping.sourceAsMap().isEmpty();
```

Comment:
Can `mapping.sourceAsMap()` be `null` here?

Reply:
in my tests, if an index does not have a mapping, it returns an empty map. the ES documentation does not specify unfortunately

Reply:
check for `mapping.sourceAsMap() != null` to be on a safe side?